urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/SealerAES.java

```diff
@@ -0,0 +1,367 @@
+	private static final int FLAG_REKEY = 0x01; // Rekey
+
+	private final SecureRandom rnd = new SecureRandom();
+	private final String cipherAlg;
+	private final int cipherKeyBits;
+	private final boolean wayAorB;
+	private final byte[] ivEncoder;
+	private final char[] secret;
+
+	private SecretKeySpec encSecretCipher;
+	private SecretKeySpec decSecretCipher;
+	private SecretKeySpec encSecretMac;
+	private SecretKeySpec decSecretMac;
+
+	private Cipher enc;
+	private Cipher dec;
+	private Mac encMac;
+	private Mac decMac;
+
+	private int encSeq = 0;
+	private int decSeq = 0;
+
+	public SealerAES(final String key, final boolean wayAorB) {
+		this(key, null, 0, wayAorB);
+	}
+
+	public SealerAES(final String key, final String cipherAlg, final int bits, final boolean wayAorB) {
+		try {
+			this.secret = key.toCharArray();
+			this.cipherAlg = (cipherAlg != null ? cipherAlg : Constants.SEALER_CIPHER_ALG);
+			this.cipherKeyBits = Math.max(bits, Constants.SEALER_CIPHER_MIN_KEY_BITS);
+			this.wayAorB = wayAorB;
+			this.enc = Cipher.getInstance(this.cipherAlg);
+			this.dec = Cipher.getInstance(this.cipherAlg);
+			this.ivEncoder = new byte[enc.getBlockSize()];
+			//
+			this.setSecret(this.secret, 0);
+		} catch (GeneralSecurityException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public String toString() {
+		return getClass().getSimpleName() + "@" + Integer.toHexString(hashCode());
+	}
+
+	public void reset() throws GeneralSecurityException {
+		Log.info(this.toString() + " reset");
+		encSeq = 0;
+		decSeq = 0;
+		setSecret(this.secret, 0);
+	}
+
+	private final void setSecret(final char[] key, final int seq) throws GeneralSecurityException {
+		Log.info(this.toString() + " cipher=" + cipherAlg + " bits=" + cipherKeyBits);
+		setSecretWayA(key, seq);
+		setSecretWayB(key, seq);
+	}
+
+	private final void setSecret(final char[] key, final int seq, final boolean wayAorB)
+			throws GeneralSecurityException {
+		if (wayAorB) {
+			setSecretWayA(key, seq);
+		} else {
+			setSecretWayB(key, seq);
+		}
+	}
+
+	private final String getAlg(final String cipherAlg) {
+		final int offset = cipherAlg.indexOf('/');
+		if (offset > 0) {
+			return cipherAlg.substring(0, offset);
+		}
+		return cipherAlg;
+	}
+
+	private final void setSecretWayA(final char[] key, final int seq) throws GeneralSecurityException {
+		final int macBitLength = Mac.getInstance(Constants.SEALER_HMAC_ALG).getMacLength() * 8;
+		final int saltSeq = miniHash(this.secret) ^ seq;
+		final SecretKeySpec secretCipherA = key(key, "cipher-A", saltSeq, 5009, cipherKeyBits,
+				getAlg(cipherAlg));
+		final SecretKeySpec secretMacA = key(key, "hmac-A", saltSeq, 5011, macBitLength,
+				Constants.SEALER_HMAC_ALG);
+		if (wayAorB) {
+			this.encSecretCipher = secretCipherA;
+			this.encSecretMac = secretMacA;
+			this.initEncMac();
+		} else {
+			this.decSecretCipher = secretCipherA;
+			this.decSecretMac = secretMacA;
+			this.initDecMac();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 100:114, 118:132.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.